Remove NOOP step

cstamas opened the pull request:
Altough JVM iface ReadWriteLock interface does not state
anything about reentrancy, let's assume that locks we use
are reentrant, and remove the NOOP step, instead always apply
given lock-step onto (now reentrant) lock.

cstamas committed:
Remove NOOP step

Altough JVM iface ReadWriteLock interface does not state
anything about reentrancy, let's assume that locks we use
are reentrant, and remove the NOOP step, instead always apply
given lock-step onto (now reentrant) lock.

michael-o commented:
This really makes sense now. Impose the hard work on the lock now, not on us. Do we still need to keep track fo the steps? I know you mentioned performance, etc. As far as I can see it is need to differentiate between lock modes, isn't it?

michael-o commented:
Also as far as I can see, a lock downgrade isn't really supported here, is it?

cstamas commented:
> This really makes sense now. Impose the hard work on the lock now, not on us. Do we still need to keep track fo the steps? I know you mentioned performance, etc. As far as I can see it is need to differentiate between lock modes, isn't it?

Yes, steps are still needed to map unlock calls onto proper (read or write) lock due how we unlock with `NamedLock.unlock` method (so as you say, to differentiate lock to be unlocked).

cstamas commented:
> Also as far as I can see, a lock downgrade isn't really supported here, is ist?

You are right, these locks support "boxed" (nested) locking. Lock downgrade assumes boxes are not nested but only overlapping... also, locking downgrade, while doable with semaphores as well (return ALL - 1 perm), would need somehow to be expressed via API, just like it happens with ReentrantReadWriteLock as well (a well defined series of calls: has exclusive, lock shared, unlock exclusive).

cstamas committed:
Remove "adapter" and use ReadWriteLock directly

As for our current two implementations (JVM and redisson)
this works just fine, no need for extra fluff

cstamas commented:
@michael-o removed whole adapter for RWLock and using ReadWriteLock iface directly. Basically if we want to go this direction, we don't need all the "adapter suff" at all (was probably biased by semaphore, that DO need it, as JVM semaphore and HZ semaphore does not share common ancestor)

michael-o commented:
> 
> 
> @michael-o removed whole adapter for RWLock and using ReadWriteLock iface directly. Basically if we want to go this direction, we don't need all the "adapter suff" at all (was probably biased by semaphore, that DO need it, as JVM semaphore and HZ semaphore does not share common ancestor)

Simpllification looks good. Though, I cannot judge on the semaphore because I have very little experience. I will leave the decision up to you. Reconciling two different approach with one adapting interface is a decent decision.
